contributors/design-proposals/federated-placement-policy.md

```diff
@@ -0,0 +1,271 @@
+existing annotations on the resource. If there are overlapping keys, the value
+defined by policy is chosen.
+
+#### Example Policy Engine Query:
+
+```http
+POST /v1/data/io/k8s/federation/annotations HTTP/1.1
+Content-Type: application/json
+```
+
+```json
+{
+  "input": {
+    "apiVersion": "extensions/v1beta1",
+    "kind": "ReplicaSet",
+    "metadata": {
+      "annotations": {
+        "eu-jurisdiction-required": "true",
+        "pci-compliance-level": "2"
+      },
+      "creationTimestamp": "2017-01-23T16:25:14Z",
+      "generation": 1,
+      "labels": {
+        "app": "nginx-eu"
+      },
+      "name": "nginx-eu",
+      "namespace": "default",
+      "resourceVersion": "364993",
+      "selfLink": "/apis/extensions/v1beta1/namespaces/default/replicasets/nginx-eu",
+      "uid": "84fab96d-e188-11e6-ac83-0a580a54020e"
+    },
+    "spec": {
+      "replicas": 4,
+      "selector": {...},
+      "template": {...},
+    }
+  }
+}
+```
+
+#### Example Policy Engine Response:
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+
+```json
+{
+  "result": {
+    "federation.kubernetes.io/replica-set-preferences": {
+      "clusters": {
+        "gce-europe-west1": {
+          "weight": 1
+        },
+        "gce-europe-west2": {
+          "weight": 1
+        }
+      },
+      "rebalance": true
+    }
+  }
+}
+```
+
+> This example shows the policy engine returning the replica-set-preferences.
+> The policy engine could similarly return a desired value for other annotations
+> such as the Cluster Selector annotation.
+
+#### Error Handling
+
+The admission controller is designed to **fail closed**. Request handling may
+fail because of:
+
+- Serialization errors
+- Request timeouts or other network errors
+- Unexpected errors from the policy engine
+
+In the event of request timeouts (or other network errors) or back-pressure
+hints from the policy engine, the admission controller should retry after
+applying a backoff.
+
+If the administrator has not defined a policy yet, the response from the policy
+engine indicates no annotations are defined. The admission controller does not
+treat this as an error, i.e., it admits the request.
+
+#### Configuration
+
+The admission controller requires configuration for the OPA endpoint:
+
+```
+{
+  “opa”: {
+    “baseURL”: “http://opa.federation.svc.cluster.local:8181/v1”,
+    “annotationsPath”: “/data/io/k8s/federation/annotations”,
+  }
+}
+```
+
+- `baseURL` specifies the URL to reach the policy engine.
+- `annotationsPath` specifies the path of the document that defines desired
+  values for annotations.
+
+The configuration file is provided to the federation-apiserver with the
+`--admission-control-config-file` command line argument.
+
+The admission controller is enabled in the federation-apiserver by providing the
+`--admission-control` command line argument. E.g.,
+`--admission-control=AlwaysAdmit,OPA`.
```

Comment:
The background here is that the admission controller written for the POC is able to apply arbitrary annotations (as well as enforce a boolean allow/deny value from the response)--the difference in the complexity of the admission controller is pretty minimal. Looking ahead, it seems valuable to be able to enforce more than just scheduling policy.

However, I'm happy to keep the proposal scoped to just scheduling policy. I've updated the proposal to use your suggested name.